datashader/reductions.py

```diff
@@ -314,6 +327,59 @@ def finalize(bases, **kwargs):
         return finalize
 
 
+class count_values(Reduction):
+    """Count of all elements in ``column``, grouped by value.
+
+    Parameters
+    ----------
+    column : str
+        Name of the column to aggregate over. Resulting aggregate has a outer dimension axis along
+        the categories present.
+    categories : array-like
+        Categories (unique values) included in the column. Any unprovided categories will be ignored
+        in the result.
+    ordered : bool
+        Treat the categories as ordered
+    """
+
+    def __init__(self, column, categories, ordered=False):
+        super(count_values, self).__init__(column=column)
+        self.categories = categories
+        self.dtype = CategoricalDtype(self.categories, ordered=ordered)
+
+    def validate(self, in_dshape):
+        if not self.column in in_dshape.dict:
+            raise ValueError("specified column not found")
+
+    def out_dshape(self, input_dshape):
+        return dshape(Record([(c, ct.int32) for c in self.dtype.categories]))
+
+    @property
+    def inputs(self):
+        return (value_codes(self.column, self.dtype), )
+
+    def _build_create(self, out_dshape):
+        n_cats = len(out_dshape.measure.fields)
+        return lambda shape: np.zeros(shape + (n_cats,), dtype='i4')
+
+    @staticmethod
+    @ngjit
+    def _append(x, y, agg, field):
+        agg[y, x, field] += 1
+
+    @staticmethod
+    def _combine(aggs):
+        return aggs.sum(axis=0, dtype='i4')
+
+    def _build_finalize(self, dshape):
+        cats = list(self.dtype.categories)
+        def finalize(bases, **kwargs):
+            dims = kwargs['dims'] + [self.column]
+            coords = kwargs['coords'] + [cats]
+            return xr.DataArray(bases[0], dims=dims, coords=coords)
+        return finalize
```

Comment:
This all seems ok, but as noted in https://github.com/bokeh/datashader/issues/140 , count_cat is unnecessarily limited to "count"; it should instead be applying any of the other reductions, but per category.  We've never implemented that functionality, but it would be a shame to make a special-purpose spark-specific version of count_cat when it could just as easily be solving the general case.  If you want to do that, there would need to be a separate PR to generalize count_cat for the existing backends, when I could merge and then here there could be a spark-specific version.

Reply:
This is an anti-pattern to work around the validation in `count_cat`. PySpark doesn't have a categorical type, so I had to do something to expose creation of a categorical column. Given the ability to cast to categorical, the implementation is the same as Pandas.